### PR TITLE
fix(observability): scope filters per surface and wire transaction/request id in logs search

### PR DESCRIPTION
**Pull request by araffass**
##  Issue 
  https://gravitee.atlassian.net/browse/APIM-14828
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
  On the Observability dashboards, the Transaction ID filter was broken on both surfaces (APIM-14828):                                                                                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                                                                                                                                                                 
  1. Overview (analytics) — the filter was offered in the Add-filter dialog, but applying it failed with Filter 'TRANSACTION_ID' is not supported for analytics queries (the analytics engine rejects it by design).                                                                                                                                                                                                             
  2. Logs — the filter was accepted but had no effect: the page's search-param builder only copied a fixed set of filter keys, and TRANSACTION_ID (and REQUEST_ID, same family) was silently dropped before the request was built — even though the whole backend chain (logs use case → CRUD → Elasticsearch adapter, v2 transaction/v4 transaction-id field mapping) already supported it.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                                                                                                                                 
##  Fix
                                                                                                                                                                                                                                                                                                                                                                                                                                     
  Per-filter surface signals (backend-derived, no drift possible). The observability filter catalog serves two surfaces; each filter now advertises which ones support it:                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                                                                                                                                                                 
  - New core enum FilterSignal { ANALYTICS, LOGS }; FilterSpec gains a signals list (compat constructor keeps existing call sites untouched).                                                                                                                                                                                                                                                                                    
  - GetAnalyticsFilterDefinitionsUseCase derives signals from the engines' own definitions: AnalyticsQueryValidator.supportsAnalytics(...) (the exact set the validator rejects — now exposed and reused) and the logs engine's FilterName enum (name-matched; HTTP_PATH ↔ URI translation documented). The 4 logs-only filters (TRANSACTION_ID, REQUEST_ID, PAYLOAD, ERROR_KEY) come out as [LOGS].                             
  - Exposed via a new optional signals property in openapi-analytics.yaml (additive; omitted when null on the sibling per-metric endpoints).                                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                                                                                                                                 
  Frontend gate per surface. New OBSERVABILITY_FILTER_SIGNAL injection token; ObservabilityFiltersApiService.getDefinitions() drops definitions not supporting the provided surface. Overview (dashboard-viewer) provides ANALYTICS, the Logs page provides LOGS. Definitions without signals (older backend during rolling upgrade) stay visible everywhere; an empty signals list means excluded. The shared dialog needs no   
  change — its existing "Logs only" badge now lights up on the Logs page.                                                                                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                                                                                                                                 
  URL-hydration prune. Conditions restored from a bookmarked/shared URL could still smuggle a logs-only filter onto Overview and reproduce the error — DashboardFiltersStore now prunes hydrated conditions whose filter isn't offered on the surface, then the page re-queries cleanly.                                                                                                                                         
                                                                                                                                                                                                                                                                                                                                                                                                                                 
  Logs wiring. buildSearchParam now passes TRANSACTION_ID and REQUEST_ID (all values — new plural transactionIds/requestIds params serialized as IN, so multi-value filters return results for every value; the scalar requestId deep-link path is untouched).

### CODE REVIEW

File: gravitee-apim-console-webui/src/management/observability/env-logs/env-logs.component.ts [hunk 1]
```diff
@@ -246,6 +252,8 @@ export class EnvLogsComponent {
       apiProductIds: filterMap.get('API_PRODUCT'),
       uri: filterMap.get('HTTP_PATH')?.[0],
```
Comment (paulatulis) [inline gravitee-apim-console-webui/src/management/observability/env-logs/env-logs.component.ts:253]: **MAJOR** — Multi-value HTTP Path on Logs silently filters by the first value only.

`supportsLogs` grants the `LOGS` signal to `HTTP_PATH` unconditionally, so the Logs Add-filter dialog now formally advertises it with its catalog operators `[EQ, IN]`. But the logs pipeline only handles it as a single-value `EQ`: this line sends `?.[0]` as the scalar `uri`, and `SearchEnvironmentLogsUseCase` explicitly ignores `ArrayFilter` for `URI`.

The path is reachable: `DashboardFiltersStore.add()` merges two conditions on the same field into one and promotes the operator via `promoteMembershipOperator` once `values.length >= 2`. So adding `HTTP Path = /a` then `HTTP Path = /b` on the Logs page yields a single `HTTP_PATH IN ['/a','/b']` condition, and only `/a` reaches the backend — no error, no warning.

The `?.[0]` truncation predates this PR, but the new signal derivation is what now asserts "the logs engine supports this filter" while only half its advertised operators actually work there. That's the same silent-drop failure class APIM-14828 is fixing. No test covers a multi-value HTTP_PATH condition.

File: gravitee-apim-console-webui/src/management/observability/dashboards/ui/dashboard-viewer/dashboard-filters.store.ts [hunk 2]
```diff
@@ -226,6 +229,21 @@ export class DashboardFiltersStore {
...
+    // engine that rejects them (APIM-14828).
+    this.definitionsPruneSubscription?.unsubscribe();
+    if (conditions.length > 0 && this.definitionProvider) {
+      this.definitionsPruneSubscription = this.definitionProvider
+        .getDefinitions()
+        .pipe(takeUntilDestroyed(this.destroyRef))
+        .subscribe(definitions => {
+          if (generation !== this.labelResolutionGeneration) return;
+          const offered = new Set(definitions.map(definition => definition.name as string));
+          this._conditions.update(current => current.filter(condition => offered.has(condition.field as string)));
```
Comment (paulatulis) [inline gravitee-apim-console-webui/src/management/observability/dashboards/ui/dashboard-viewer/dashboard-filters.store.ts]: **MAJOR** — The prune re-queries on every bookmarked-filter load, including when nothing is pruned.

`current.filter(...)` returns a new array reference even when no condition is removed, and Angular signals compare by reference. So on every hydration that has *any* URL filters — including entirely valid ones — the async arrival of definitions re-fires `_conditions`, `requestFilters` recomputes, and every consumer re-queries.

On env-logs the `searchParams -> toObservable -> switchMap` chain cancels the in-flight initial fetch and restarts it, roughly doubling initial load latency for every bookmarked-filter visit. On the dashboard, the `[requestFilters]` input change re-triggers all widget queries. This happens on the path where no pruning was needed at all.

Comment (paulatulis) [inline gravitee-apim-console-webui/src/management/observability/dashboards/ui/dashboard-viewer/dashboard-filters.store.ts]: **MAJOR** — When the prune does drop a condition, the URL still carries it.

`syncToRouter()` is never called here, so the `q` query param keeps encoding the pruned filter while the chips and queries no longer include it. Three consequences:

1. The visible state and the shareable URL disagree.
2. The prune is asynchronous, so on every reload the initial query still fires with the unsupported filter *before* definitions arrive — `AnalyticsQueryValidator` rejects it and the analytics widgets flash the error state, then the prune runs and re-queries. The error the ticket is about is still reachable, just transiently.
3. Re-sharing the still-dirty URL keeps propagating the unsupported filter.

So AC4 ("a bookmarked/shared URL must not reintroduce an unsupported filter") looks only half-resolved.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-analytics.yaml [hunk 3]
```diff
@@ -1660,12 +1660,26 @@ components:
                         $ref: "#/components/schemas/ApiName"
                     description: List of API types this filter applies to, derived from the metrics that reference it.
                     example: ["HTTP_PROXY", "LLM"]
+                signals:
```
Comment (paulatulis) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-analytics.yaml:1663]: **MAJOR** — `signals` is advertised inconsistently across the two endpoints that share this schema, and the contract isn't documented.

`FilterSpecsResponse` is referenced by two operations: `queryObservabilityFilterDefinitions` (line ~259) and `queryFilterSpecsForMetric` (line ~150, via `GetMetricFilterSpecUseCase` and the same `AnalyticsDefinitionMapper.toFilterSpecsResponse`). Signals are derived only in `GetAnalyticsFilterDefinitionsUseCase`, so the identical catalog filter (e.g. `API`) comes back **with** `signals` from one endpoint and `signals: null` from the other.

Meanwhile this description documents the property unconditionally and says nothing about absence or emptiness — yet the console gives those cases opposite meanings: absent = "no restriction, show everywhere" (the rolling-upgrade path), empty array = "no surface supports it, exclude". A consumer or codegen client reading only the spec can't learn that absence is expected on the per-metric endpoint, or that an empty array is meaningful and exclusionary.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCase.java [hunk 4]
```diff
@@ -32,6 +43,39 @@ public GetAnalyticsFilterDefinitionsUseCase(AnalyticsDefinitionQueryService defi
...
+        if (AnalyticsQueryValidator.supportsAnalytics(name)) {
+            signals.add(FilterSignal.ANALYTICS);
+        }
+        if (supportsLogs(name)) {
+            signals.add(FilterSignal.LOGS);
+        }
+        return List.copyOf(signals);
+    }
+
+    private static boolean supportsLogs(FilterSpec.Name name) {
```
Comment (paulatulis) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCase.java:73]: **MAJOR** — Surface support is derived by matching two independently-evolving enums by name, with no compile-time or test guard.

`supportsLogs` matches `FilterSpec.Name.name()` against `FilterName.values()`, plus a hard-coded `HTTP_PATH` -> `URI` special case and a comment noting two more (`MCP_METHOD`, `RESPONSE_TIME`) that silently fall through unmapped. Renaming or adding a value in either enum silently changes which filters appear on which surface — nothing fails to compile and no test catches it. Renaming `FilterName.TRANSACTION_ID`, for instance, would quietly remove the filter from the Logs page.

Zooming out, "which surface supports which filter" is now encoded in four independent places that can drift: this name-matching, `AnalyticsQueryValidator.ANALYTICS_UNSUPPORTED_FILTERS` (4 names), `AnalyticsDefinitionMapper`'s `@ValueMapping(source = "PAYLOAD", target = THROW_EXCEPTION)` guard, and the per-metric `filters:` lists in `analytics-definition.yaml`. The last two already disagree with the validator — the mapper hard-fails only on `PAYLOAD` while the validator rejects four names — so adding `TRANSACTION_ID`, `REQUEST_ID`, or `ERROR_KEY` to any metric's yaml filter list would make the per-metric endpoint advertise a filter the analytics engine rejects. That's this exact bug class, one yaml edit away.

(Related: an `analytics_engine` use case now imports `logs_engine.model.FilterName` directly, with no port abstracting the concern.)

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/model/FilterSpec.java [hunk 5]
```diff
@@ -27,8 +28,29 @@ public record FilterSpec(
     List<String> enumValues,
```
Comment (paulatulis) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/model/FilterSpec.java:33]: **MAJOR** — `FilterSpec` is now an 8-component record with a telescoping compatibility constructor.

The record gains an 8th component, a 7-arg compat constructor, and a hand-written `withSignals` wither. `java-conventions.md` ("Records and Builder Pattern") says: *"once a type has more than ~5 fields, positional constructors become fragile in tests and `with*` copy-methods multiply with every new field. Prefer a `@Builder`-annotated class (Lombok) or a companion builder."*

The next component added will force a third constructor overload and another wither, and tests already construct specs with long positional `null` runs.

Acknowledging this is a refactor ask on a targeted bugfix — reasonable to defer, but flagging it as a real deviation rather than let it pass silently.

File: gravitee-apim-console-webui/src/management/observability/dashboards/ui/dashboard-viewer/dashboard-filters.store.spec.ts [hunk 6]
```diff
@@ -316,6 +316,45 @@ describe('DashboardFiltersStore', () => {
     expect(hydratedStore.periodControl.value.to!.valueOf()).toBe(2000000);
   });
 
+  it('should drop URL-hydrated conditions for filters the surface does not offer', () => {
```
Comment (paulatulis) [inline gravitee-apim-console-webui/src/management/observability/dashboards/ui/dashboard-viewer/dashboard-filters.store.spec.ts]: **MAJOR** — This test can't exercise the ordering window the prune actually has to survive.

The `FILTER_DEFINITION_PROVIDER` mock returns `of([...])`, which emits synchronously — so by the time the assertion runs, definitions have always already arrived. The real path is the opposite: definitions have *not* loaded when the URL is hydrated, and conditions must stay visible until they arrive and be pruned only on emission.

That window is exactly where the transient error-flash behaviour lives (the initial query fires with the unsupported filter before the prune can run), so it's the case most worth pinning.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCaseTest.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+import io.gravitee.apim.core.observability.model.FilterType;
+import java.util.List;
+import org.junit.jupiter.api.DisplayNameGeneration;
+import org.junit.jupiter.api.DisplayNameGenerator;
+import org.junit.jupiter.api.Test;
+
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+class GetAnalyticsFilterDefinitionsUseCaseTest {
+
+    private final AnalyticsDefinitionQueryService definition = mock(AnalyticsDefinitionQueryService.class);
```
Comment (paulatulis) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCaseTest.java:43]: **MAJOR (low confidence)** — This mocks the `AnalyticsDefinitionQueryService` port; the testing conventions call for in-memory port fakes rather than Mockito mocks on ports.

Flagging with a caveat: no `InMemoryAnalyticsDefinitionQueryService` exists yet, and sibling use-case tests in this module use the same Mockito pattern. If that convention isn't actually enforced here, treat this as noise — it's a maintainer's call, not something I'd block on.

File: gravitee-apim-console-webui/src/management/observability/env-logs/env-logs.component.ts [hunk 8]
```diff
@@ -246,6 +252,8 @@ export class EnvLogsComponent {
...
       bodyText: filterMap.get('PAYLOAD')?.[0],
```
Comment (paulatulis) [inline gravitee-apim-console-webui/src/management/observability/env-logs/env-logs.component.ts:254]: **MINOR** — The `PAYLOAD` -> `bodyText` mapping has no component-level wiring test.

Every other mapped field in the `filter wiring` describe block has one (API, PLAN, HTTP_PATH, TRANSACTION_ID, REQUEST_ID, ERROR_KEY, ...). PAYLOAD is the only mapped field left unexercised, so a rename or typo in the map key would ship silently.

File: gravitee-apim-console-webui/src/services-ngx/environment-logs.service.ts [hunk 9]
```diff
@@ -118,6 +122,8 @@ function buildFilters(param?: SearchLogsParam): LogFilter[] {
     { name: 'ENTRYPOINT', values: param.entrypoints },
     { name: 'ERROR_KEY', values: param.errorKeys },
     { name: 'API_PRODUCT', values: param.apiProductIds },
+    { name: 'REQUEST_ID', values: param.requestIds },
```
Comment (paulatulis) [inline gravitee-apim-console-webui/src/services-ngx/environment-logs.service.ts:125]: **MINOR** — The scalar and plural variants of the same id can both be populated, and precedence is undefined.

`buildFilters()` builds `scalarFilters` from `requestId`/`transactionId` (`EQ`) and `arrayFilters` from `requestIds`/`transactionIds` (`IN`) independently, then concatenates both into the same `filters[]` array with no mutual-exclusion check. A caller populating both would emit two entries with the same `name` but different operators, and `openapi-logs.yaml` says nothing about how the backend resolves duplicate same-name filter entries.

Latent today — no caller sets both, and the scalar deep-link path is untouched — but nothing pins the behaviour either way.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/model/FilterSpec.java [hunk 10]
```diff
@@ -27,8 +28,29 @@ public record FilterSpec(
...
     NumberRange range,
     List<FilterOperator> operators,
-    List<ApiSpec.Name> apis
+    List<ApiSpec.Name> apis,
+    List<FilterSignal> signals
```
Comment (paulatulis) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/model/FilterSpec.java:37]: **MINOR** — The new `signals` component isn't defensively copied, so a caller-held mutable list can be mutated after construction, silently changing record equality and serialized output.

The sole current producer happens to pass `List.copyOf(...)`, and the pre-existing list components (`enumValues`, `operators`, `apis`) share the same gap — so this is a consistency nit rather than a live defect.

Comment (paulatulis) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/model/FilterSpec.java:37]: **MINOR** — The null-vs-empty contract for `signals` is load-bearing but documented nowhere on the record.

Null means "signals were not derived, no gating applies" and an empty list means "supported by no surface, hide everywhere" — the frontend gate depends on that distinction. The compat-constructor Javadoc only says signals are derived elsewhere, and the OpenAPI description omits it too.

Empty-as-"excluded everywhere" is a surprising sentinel that a future catalog author could trip over without ever hitting a test failure.

File: gravitee-apim-console-webui/src/management/observability/dashboards/ui/dashboard-viewer/dashboard-filters.store.ts [hunk 11]
```diff
@@ -226,6 +229,21 @@ export class DashboardFiltersStore {
...
         });
     }
+
+    // Conditions restored from the URL may reference filters this surface does not offer (e.g. a
+    // logs-only filter deep-linked onto an analytics dashboard) — drop them instead of querying an
+    // engine that rejects them (APIM-14828).
+    this.definitionsPruneSubscription?.unsubscribe();
+    if (conditions.length > 0 && this.definitionProvider) {
+      this.definitionsPruneSubscription = this.definitionProvider
+        .getDefinitions()
```
Comment (paulatulis) [inline gravitee-apim-console-webui/src/management/observability/dashboards/ui/dashboard-viewer/dashboard-filters.store.ts:241]: **MINOR** — Extra definitions fetch per hydration, and a failed fetch is invisible.

This subscribes to a fresh `getDefinitions()` HTTP request on every hydration, and `hydrateFromRoute` runs twice at startup (constructor snapshot plus the first synchronous emission of `route.queryParams`), plus again on any unrelated query-param change — each issuing another GET of `/observability/filters/definition` on top of the filter bar's own call. The service doesn't cache.

There's also no error handler: if the definitions request fails, it surfaces only as an unhandled RxJS error and the prune silently never applies.

File: gravitee-apim-console-webui/src/management/observability/data-access/observability-filters-api.service.ts [hunk 12]
```diff
@@ -63,14 +63,26 @@ interface FilterValueItemApi {
...
 }
 
+/** Observability surface consuming the shared filter catalog — mirrors the backend FilterSignal enum. */
+export type ObservabilityFilterSignal = 'ANALYTICS' | 'LOGS';
+
+/**
+ * Surface hosting the filter bar. When provided, definitions carrying signals are narrowed to that
+ * surface so a page only offers filters its query engine supports (APIM-14828).
+ */
+export const OBSERVABILITY_FILTER_SIGNAL = new InjectionToken<ObservabilityFilterSignal>('OBSERVABILITY_FILTER_SIGNAL');
```
Comment (paulatulis) [inline gravitee-apim-console-webui/src/management/observability/data-access/observability-filters-api.service.ts:73]: **MINOR** — Design observation on the ambient token.

`OBSERVABILITY_FILTER_SIGNAL` is injected `{ optional: true }` inside a shared `data-access` service, so `getDefinitions()` silently branches on whatever the enclosing feature component happened to provide. Both current consumers (dashboard-viewer, env-logs) do provide it, so nothing is broken today.

The concern is that a future consumer of `ObservabilityFiltersApiService` that forgets the token gets the ungated legacy behaviour with no compiler warning and no test failure — the failure mode is silent, which is what makes it worth noting.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCaseTest.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+            )
+        );
+
+        var specs = useCase.execute().specs();
+
+        assertThat(specs)
+            .extracting(FilterSpec::name, FilterSpec::signals)
+            .containsExactly(
+                tuple(FilterSpec.Name.API, List.of(FilterSignal.ANALYTICS, FilterSignal.LOGS)),
+                // Logs-only: rejected by the analytics engine (APIM-14828).
```
Comment (paulatulis) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCaseTest.java]: **MINOR** — Ticket reference in a test-body comment. The conventions permit the explanatory "why" prose but not the embedded ticket tag.

```suggestion
                // Logs-only: rejected by the analytics engine.
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/observability/ObservabilityFiltersDefinitionResourceTest.java [hunk 14]
```diff
@@ -72,6 +74,28 @@ void should_return_filter_definitions() {
...
+    @Test
+    void should_advertise_filter_signals_per_surface() {
+        var response = rootTarget().request().get();
+
+        assertThat(response)
+            .hasStatus(200)
+            .asEntity(FilterSpecsResponse.class)
+            .extracting(FilterSpecsResponse::getData)
+            .satisfies(filters -> {
+                // Logs-only filters must not be offered on the analytics surface (APIM-14828).
```
Comment (paulatulis) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/observability/ObservabilityFiltersDefinitionResourceTest.java]: **MINOR** — Ticket reference in a test-body comment. The conventions permit the explanatory "why" prose but not the embedded ticket tag.

```suggestion
                // Logs-only filters must not be offered on the analytics surface.
```

File: gravitee-apim-console-webui/src/management/observability/env-logs/env-logs.component.ts (hunk 1 above)
Comment (araffass) [reply gravitee-apim-console-webui/src/management/observability/env-logs/env-logs.component.ts:253]: Legit, but the backend uri filter is a wildcard query and single value by design, so real IN support means an or-of-wildcards across the model, crud and the ES adapter. Out of scope here. Keeping the LOGS signal since EQ works today, removing it would drop the filter from the logs page entirely. Opening a follow up ticket for multi value uri.

File: gravitee-apim-console-webui/src/management/observability/dashboards/ui/dashboard-viewer/dashboard-filters.store.ts (hunk 2 above)
Comment (araffass) [reply gravitee-apim-console-webui/src/management/observability/dashboards/ui/dashboard-viewer/dashboard-filters.store.ts]: Good catch, fixed. The prune keeps the same array reference when nothing is dropped, added a test asserting it.

Comment (araffass) [reply gravitee-apim-console-webui/src/management/observability/dashboards/ui/dashboard-viewer/dashboard-filters.store.ts]: Fixed the URL part, syncToRouter now runs when something is actually pruned. The transient first query before definitions arrive is still possible, killing it completely would mean holding every initial query on the catalog fetch which felt worse. Documented in the code, it self heals after the prune.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-analytics.yaml (hunk 3 above)
Comment (araffass) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-analytics.yaml:1663]: Fixed. The description now says it is only populated by the observability definitions operation, absent means no restriction and empty means offered nowhere. Same contract documented on the core record.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCase.java (hunk 4 above)
Comment (araffass) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCase.java:73]: Added a guard test: every logs FilterName must be a catalog name, the URI translation or a documented exception, otherwise it fails pointing at supportsLogs. Not compile time safe but drift can no longer ship silently.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/model/FilterSpec.java (hunk 5 above)
Comment (araffass) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/model/FilterSpec.java:33]: Agreed on the direction, keeping it out of this bugfix as you suggested. Will do it when the next field shows up.

File: gravitee-apim-console-webui/src/management/observability/dashboards/ui/dashboard-viewer/dashboard-filters.store.spec.ts (hunk 6 above)
Comment (araffass) [reply gravitee-apim-console-webui/src/management/observability/dashboards/ui/dashboard-viewer/dashboard-filters.store.spec.ts]: Right, rewritten with a Subject. It now asserts conditions are still visible before the emission and pruned after, plus the router sync.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCaseTest.java (hunk 7 above)
Comment (araffass) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCaseTest.java:43]: Kept the mock, sibling use case tests in this module do the same and there is no in-memory fake for this port yet.

File: gravitee-apim-console-webui/src/management/observability/env-logs/env-logs.component.ts (hunk 8 above)
Comment (araffass) [reply gravitee-apim-console-webui/src/management/observability/env-logs/env-logs.component.ts:254]: Added.

File: gravitee-apim-console-webui/src/services-ngx/environment-logs.service.ts (hunk 9 above)
Comment (araffass) [reply gravitee-apim-console-webui/src/services-ngx/environment-logs.service.ts:125]: Fixed, the plural variant wins and the scalar is skipped, with a test.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/model/FilterSpec.java (hunk 10 above)
Comment (araffass) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/model/FilterSpec.java:37]: Leaving as is to stay consistent with the other list components. Happy to do a List.copyOf pass on the whole record separately.

Comment (araffass) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/model/FilterSpec.java:37]: Documented on the record javadoc and in the openapi description.

File: gravitee-apim-console-webui/src/management/observability/dashboards/ui/dashboard-viewer/dashboard-filters.store.ts (hunk 11 above)
Comment (araffass) [reply gravitee-apim-console-webui/src/management/observability/dashboards/ui/dashboard-viewer/dashboard-filters.store.ts:241]: Added a catchError so a failed fetch keeps the restored filters instead of failing silently. Left the extra GET, the endpoint is cheap and caching it in the service would change freshness for the add filter dialog.

File: gravitee-apim-console-webui/src/management/observability/data-access/observability-filters-api.service.ts (hunk 12 above)
Comment (araffass) [reply gravitee-apim-console-webui/src/management/observability/data-access/observability-filters-api.service.ts:73]: Fair point, it is the tradeoff of the optional token. The two current pages now pin their token value with a test, so at least forgetting it on an existing page fails.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCaseTest.java (hunk 13 above)
Comment (araffass) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCaseTest.java]: Removed.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/observability/ObservabilityFiltersDefinitionResourceTest.java (hunk 14 above)
Comment (araffass) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/observability/ObservabilityFiltersDefinitionResourceTest.java]: Removed.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCase.java (hunk 4 above)
Comment (paulatulis) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCase.java:73]: AnalyticsDefinitionMapper still carries only @ValueMapping(source = "PAYLOAD", target = THROW_EXCEPTION) while AnalyticsQueryValidator rejects four names (PAYLOAD, ERROR_KEY, REQUEST_ID, TRANSACTION_ID). So adding TRANSACTION_ID to any metric's yaml filter list still makes the per-metric endpoint advertise a filter the analytics engine rejects, with nothing to catch it. That was the "one yaml edit away" hazard, and it stands.

**Review by paulatulis — APPROVED** [review]
Left follow up comment to one thing. I will let you decide if you want to address it. Otherwise LGTM!
https://github.com/gravitee-io/gravitee-api-management/pull/18864#discussion_r3691371144

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCase.java (hunk 4 above)
Comment (araffass) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/analytics_engine/use_case/GetAnalyticsFilterDefinitionsUseCase.java:73]: Fair, the guard I added only covered the logs axis. Now added a test that walks every metric's yaml filter list through the real definition and fails on any analytics rejected name, so the one yaml edit away case is a build failure. Also aligned the mapper's throw list with the validator's four names so runtime stays consistent if anything bypasses the yaml path.
